### PR TITLE
Add Docker task with Artifactory

### PR DESCRIPTION
**Pull request by barbelity**

### CODE REVIEW

**Comment by eyalbe4** [comment]
@barbelity,
In addition to my inline comments, we'll probably need to add an Artifactory Repository flag to the docker  tests. This is because the repository name used by the test will have to match the configured docker registry domain.

File: tasks/ArtifactoryDocker/task.json [hunk 1]
```diff
@@ -0,0 +1,88 @@
+{
+    "id": "07fc5c20-6882-a1c2-cbcd-6531d5a72113",
+    "name": "ArtifactoryDocker",
+    "friendlyName": "Artifactory Docker",
+    "description": "Publish docker images to Artifactory, while allowing to collect build-info. The collected build-info can be later published to Artifactory by the \"Artifactory Publish Build Info\" task.",
```
[the remark this answers is not in the document]
Comment (barbelity) [reply tasks/ArtifactoryDocker/task.json]: fixed

File: tasks/ArtifactoryDocker/docker.js [hunk 2]
```diff
@@ -0,0 +1,51 @@
+
```
[the remark this answers is not in the document]
Comment (barbelity) [reply tasks/ArtifactoryDocker/docker.js]: fixed

File: tasks/ArtifactoryDocker/dockerBuild.js [hunk 3]
```diff
@@ -0,0 +1,51 @@
+
+const tl = require('vsts-task-lib/task');
+const utils = require('artifactory-tasks-utils');
+
+const dockerPushCommand = "rt dp";
+
+function RunTaskCbk(cliPath) {
+    // Validate docker exists on agent
+    tl.which("docker", true);
```
Comment (yahavi) [inline tasks/ArtifactoryDocker/dockerBuild.js]: Please don't leave unhandled exceptions. Thanks!

File: tasks/ArtifactoryDocker/docker.js [hunk 4]
```diff
@@ -0,0 +1,51 @@
...
+    let targetRepository = tl.getInput("targetRepo", true);
+    let imageTag = tl.getInput("imageTag", true);
+
+    // Build the cli command
+    let cliCommand = utils.cliJoin(cliPath, dockerPushCommand, utils.quote(imageTag), utils.quote(targetRepository), "--url=" + utils.quote(artifactoryUrl));
+    cliCommand = utils.addArtifactoryCredentials(cliCommand, artifactoryService);
+
+    // Add build info collection
+    if (collectBuildInfo) {
+        cliCommand = utils.cliJoin(cliCommand, "--build-name=" + utils.quote(buildDefinition), "--build-number=" + utils.quote(buildNumber));
```
[the remark this answers is not in the document]
Comment (barbelity) [reply tasks/ArtifactoryDocker/docker.js]: Will be changed at #25

File: tests/tests.js [hunk 5]
```diff
@@ -238,6 +244,26 @@ describe("JFrog Artifactory VSTS Extension Tests", () => {
...
+                // Run docker build + tag
+                execSync("docker build -t " + testUtils.artiactoryDockerTag + " " + path.join(__dirname, "resources", testDir, filesDir));
+                // run docker push
+                mockTask(testDir, "push");
+                mockTask(testDir, "publish");
+                getAndAssertBuild("dockerTest", "1");
+                deleteBuild("dockerTest");
+            })
+        } else {
+            console.log("Could not find 'docker' executable, skipping Docker tests");
```
[the remark this answers is not in the document]
Comment (barbelity) [reply tests/tests.js]: Will be added in a separated PR.

File: tasks/ArtifactoryDocker/docker.js [hunk 6]
```diff
@@ -0,0 +1,51 @@
...
+    tl.which("docker", true);
+
+    let defaultWorkDir = tl.getVariable('System.DefaultWorkingDirectory');
+    if (!defaultWorkDir) {
+        tl.setResult(tl.TaskResult.Failed, "Failed getting default working directory.");
+        return;
+    }
+    let buildDefinition = tl.getVariable('Build.DefinitionName');
+    let buildNumber = tl.getVariable('Build.BuildNumber');
+
```
[the remark this answers is not in the document]
Comment (barbelity) [reply tasks/ArtifactoryDocker/docker.js]: Will be resolved in a separated PR

**Pull request merged**